### ISSUE TITLE
Unspecified imports when imports multiple modules as the same alias

### ISSUE THREAD

**Issue by tfausak** [purescript/purescript#2227]
I sometimes want to import many modules under the same alias. For instance, I may want to pull in both [`Node.FS`](https://pursuit.purescript.org/packages/purescript-node-fs/1.0.0/docs/Node.FS) and [`Node.FS.Async`](https://pursuit.purescript.org/packages/purescript-node-fs/1.0.0/docs/Node.FS.Async) in as `FS`. 

``` purescript
import Node.FS as FS
import Node.FS.Async as FS
```

I want to do this so that I can specify the type of some action and actually perform that action with a single namespace.

``` purescript
main :: Eff (fs :: FS.FS) Unit
main = FS.rename "foo" "bar" \ _ -> pure unit
```

Doing so gave me a warning, which surprised me.

```
Module Node.FS was imported as FS with unspecified imports.
As there are multiple modules being imported as FS, consider using the explicit form:
  import Node.FS (FS) as FS

Module Node.FS.Async was imported as FS with unspecified imports.
As there are multiple modules being imported as FS, consider using the explicit form:
  import Node.FS.Async (rename) as FS
```

I'm familiar with these warnings when importing unqualified. However I wish I didn't get these warnings when importing qualified. I import qualified because I don't want to explicitly list imports. And I'm importing multiple things under the same alias because I know that they won't overlap (or that I'm okay with them overlapping).

**Comment by paf31** [comment]
Qualified imports run the same risk as unqualified ones: if you have multiple, then upstream library changes can break your code later, so we have the same warning.

However, I think we should allow one open qualified import per namespace, which we don't currently do, if I remember correctly.

**Comment by tfausak** [comment]
I thought you might say that 🙂

I feel like importing things under an alias is notably different than importing them globally unqualified. When I do this, the modules typically all come the same package. Even if they didn't, the increased usability is worth the danger of clobbering imports. 

Furthermore, PureScript's error messages are already pretty great in this case. 

``` purescript
module A where data D
module B where data D
module C where
 import A as M
 import B as M
 type T = M.D
```

```
Conflicting definitions are in scope for type D from the following modules:
  A
  B
```

And if you explicitly list `D` from `B`, you also get a great warning.

``` purescript
module C where
 import A as M
 import B (D) as M
 type T = M.D
```

```
Shadowed definitions are in scope for type D from the following open imports:
  import A
These will be ignored and the declaration from B will be used.
```

In short, I feel like the desire to avoid a pretty uncommon scenario is negatively affecting the usability of imports in PureScript for me. What is the recommended way to import and use both `Node.FS` and `Node.FS.Async`? 

``` purescript
-- Generates warnings.
import Node.FS as FS
import Node.FS.Async as FS

-- Annoying and tedious to write and update.
import Node.FS (FS, ...)
import Node.FS.Async (rename, ...)

-- Worst of all worlds.
import Node.FS (FS, ..) as FS
import Node.FS.Async (rename, ...) as FS

-- Annoying to actually use and kind of silly to write.
import Node.FS as Node.FS
import Node.FS.Async as Node.FS.Async
```

> I think we should allow one open qualified import per namespace

I think that is currently allowed.

**Comment by natefaubion** [comment]
FWIW, Halogen has some comically long reexport lists because of this. It actually makes it very easy to forget to reexport some definitions!

**Comment by paf31** [comment]
I appreciate this is a bit of a pain, but it's a pain for exactly the same reasons as before, which we've worked through already. I think we either need to admit this is a good thing in general, or come up with some other restriction. Unfortunately, saying something like "restrict qualified imports with the same name to come from the same package" isn't going to cut it, since the compiler has no knowledge of package structure.

**Comment by tfausak** [comment]
I'm arguing that pulling multiple modules in as the same name acknowledges the potential problems in a way that importing them unqualified does not. I agree that `import A; import B` is problematic. I think that `import A as M; import B as M` is *not* problematic. If there is a conflict, I explicitly opted into making that conflict happen.

**Comment by texastoland** [comment]
Not disagreeing with the motivation but IIRC the line of reasoning was that adding identifiers could still constitute a breaking change to code importing it (and recursively importing it).

**Comment by garyb** [comment]
^ exactly. It may well be that by far the majority case that importing `as M` from multiple modules is safe, but there's no guarantee that it is so in every case.

**Comment by tfausak** [comment]
Here it is, months after #1869 and I still don't really get why this one particular problem is deciding how imports work for an entire language. 

To make sure that I understand correctly, is this an accurate summary of the problem?

``` purescript
-- purescript-a@1.0.0
module A where
data A

-- purescript-b@1.0.0
module B where
data B

-- purescript-c@1.0.0
-- depends on purescript-a ^1.0.0 and purescript-b ^1.0.0
module C where
import A
import B

-- purescript-a@1.1.0
module A where
data A
data B
```

With that scenario, when someone tries to install purescript-c, they will get a compile-time error about conflicting imports. That's the problem in a nutshell, right?

I can see how this might look like I'm rehashing the same argument again, but I think the situation is different for aliased imports.

**Comment by texastoland** [comment]
> With that scenario, when someone tries to install purescript-c, they will get a compile-time error about conflicting imports. That's the problem in a nutshell, right?

Correct. More importantly if `C` is your app's dependency and `A` breaks it's a pain to figure out which dependency needs locked down to fix the inherited error.

**Comment by texastoland** [comment]
You pointed out in the previous issue that if the both `module A` and `module B` were local to your package then you could guarantee its safety. The problem was no way identify modules with their package. My impression of that issues resolution was wait to see what happens with first class modules?

**Comment by tfausak** [comment]
For this issue, I don't particularly care if the modules come from the same package or not. I don't have any PureScript examples handy, but consider importing [`Data.List`](https://www.stackage.org/haddock/lts-6.7/base-4.8.2.0/Data-List.html) and [`Data.List.Utils`](https://www.stackage.org/haddock/lts-6.7/MissingH-1.3.0.2/Data-List-Utils.html) both as `List`. I obviously can't *guarantee* that arbitrary versions of those packages can be combined without conflicts, but I also obviously want to work with them as if they were a single package. 

I think I mentioned this elsewhere, but I want to restate it here: This is a problem that is more easily solved at some other point in the build process. For instance, it could be that I specified *exact* version numbers for purescript-a and purescript-b in my `bower.json`. In that case it is not possible for anything to suddenly break. Now I'm sure you'll say that if I did that then I should just ignore the warning, but that's not satisfactory to me. I typically treat warnings and errors and try to build everything warning-free. Since `psc` doesn't allow configuring warnings, that leaves me between a rock and a hard place.

**Comment by garyb** [comment]
In the example you posted, `purescript-c@1.0.0` and every dependency of it is also broken by the release of `purescript-a@1.1.0`.

**Comment by tfausak** [comment]
Yes, that's what I meant by: "when someone tries to install purescript-c, they will get a compile-time error about conflicting imports". It's broken. Anything that depends on it is broken.

**Comment by paf31** [comment]
Right, so that's a possible future problem, but not a problem today, and that's a good fit for a warning.

**Comment by tfausak** [comment]
It may not be a possible future problem, depending on how you dependencies are specified. Heck, things that depend on purescript-c could be fine if they specified purescript-a@1.0.0 in their resolutions.

I feel like this is a problem in the same way that `import A (A)` is a problem if you've specified `"purescript-a": ">= 0"`. You're code might break in the future if you do that because the module `A` could decide to stop exporting the data type `A` (or make it a type alias or a type class or an effect).

**Comment by garyb** [comment]
Oh, for the record, as I don't think anyone answered this:

> What is the recommended way to import and use both Node.FS and Node.FS.Async?

I would import them as `FS` and `FSA`. That seems perfectly acceptable to me, as I'd like to know which thing I was referring to anyway. The main time I see this issue being a problem is in cases where a module doesn't re-export things that it could, as that's the only time I'd want to import multiple things qualified with the same name. Like:

```
import Foo.Bar.M1 as Bar
import Foo.Bar.M2 as Bar
import Foo.Bar.M3 as Bar
```

In this situation the problem is there's no `Foo.Bar` defined re-exporting the children.

**Comment by tfausak** [comment]
A real-life example of that is [`Control.Monad.Eff`](https://pursuit.purescript.org/packages/purescript-eff/1.0.0/docs/Control.Monad.Eff) and [`Control.Monad.Eff.Class`](https://pursuit.purescript.org/packages/purescript-eff/1.0.0/docs/Control.Monad.Eff.Class). I routinely want to use both `Eff.runPure` and `Eff.liftEff` instead of `EffC.liftEff` or `EffClass.liftEff` or `MonadEff.liftEff` or whatever.

**Comment by hdgarrood** [comment]
> Since psc doesn't allow configuring warnings, that leaves me between a rock and a hard place.

I think this is the root of the issue &mdash; this is of course correct, although I'd put it slightly differently. I'd say that `psc` doesn't take responsibility for filtering warnings, and instead leaves it to other tools. From this perspective, the right answer is arguably to just ignore those warnings (i.e. configure a tool like `psa` to do so).

I think this approach ought to work for application developers. However, as far as I know, there's no good way for library authors to indicate to people consuming their code that specific warnings should be ignored. I think the last time this was brought up, the conclusion was that that's probably a good thing, since library authors should be aiming for warning-free code; do we agree on this?

**Comment by garyb** [comment]
> since library authors should be aiming for warning-free code

Yes. I would be hesitant to use a library that raises warnings (now they're fixed that is ;)). Especially this warning.

**Comment by tfausak** [comment]
> However, as far as I know, there's no good way for library authors to indicate to people consuming their code that specific warnings should be ignored.

This is exactly why I insist on writing warning-free code. I don't want to generate a bunch of warnings when other people use my software, even if I am comfortable with those warnings. In Haskell I can sidestep this entire issue with `{-# OPTIONS_GHC -fno-warn-whatever #-}`, but there is no such escape hatch in PureScript.

**Comment by paf31** [comment]
I'd like to close this, or assign it to a milestone. Thoughts?

**Comment by garyb** [comment]
I vote close, it's annoying, but it's the right way to do it. The only way to avoid it is for the compiler to have package info, or give up semver and say that any non-fixed versioning may break your code. The latter is unacceptable in my opinion.

**Comment by tfausak** [comment]
I am strongly in favor of allowing this behavior without any warnings. Consider the case where you want to use many of the modules provided by the [node-fs](https://pursuit.purescript.org/packages/purescript-node-fs/1.0.0) package. You have to do this monstrosity to avoid warnings:

``` purescript
import Node.FS        as FS
import Node.FS.Async  as FSA
import Node.FS.Perms  as FSP
import Node.FS.Stats  as FSS
import Node.FS.Stream as FST -- Name collision with stats.
```

What's worse, nothing stops me from setting myself up for the same problem without any warnings at all. Consider the following module:

``` purescript
module My.FS
    ( module Node.FS
    , module Node.FS.Async
    , module Node.FS.Perms
    , module Node.FS.Stats
    , module Node.FS.Stream
    ) where
import Node.FS        as Node.FS
import Node.FS.Async  as Node.FS.Async
import Node.FS.Perms  as Node.FS.Perms
import Node.FS.Stats  as Node.FS.Stats
import Node.FS.Stream as Node.FS.Stream
```

Compiles cleanly without warnings. Allows me to `import My.FS as FS` and be on my merry way. Extremely tedious to write for every group of modules that I want to import together. 

You could argue that this is a problem with the node-fs package and that it should provide a single top-level module. I wouldn't disagree with you, but there are many packages like this. I'm not going to open a pull request every time I want to do this.

**Comment by garyb** [comment]
I just can't ever rank convenience over having code randomly break because a dependency changed.

**Comment by hdgarrood** [comment]
Also, if people are writing libraries that give extra functions to modules like Data.List, instead of doing the following:

```
import Data.List as List
import Data.List.Utils as List
```

You could instead re-export the entirety of both modules from a third, using explicit import lists to avoid any warnings (and breakage risk). This only needs to be done once if it's done inside whichever package defines Data.List.Utils.

**Comment by texastoland** [comment]
I vote to close too. I empathize with the inconvenience but care strongly about versioning having been bitten by it in npm. Moreover I think @hdgarrood and @paf31 were on the right track in #1901 suggesting an underlying issue rather than working around symptoms of it.

**Comment by paf31** [comment]
Ok, so I think we know the solution(s), even if they are inconvenient in the short term:

- Don't use overlapping qualified imports without explicit import lists, and/or
- Go through standard libraries and start re-exporting more things where appropriate.
